include/llbuild/Core/DependencyKeyIDs.h

```diff
@@ -0,0 +1,154 @@
+//===- DependencyKeyIDs.h ---------------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLBUILD_CORE_ATTRIBUTEDKEYIDS_H
+#define LLBUILD_CORE_ATTRIBUTEDKEYIDS_H
+
+#include "llbuild/Core/KeyID.h"
+
+#include <cstdint>
+#include <utility>
+#include <vector>
+
+namespace llbuild {
+namespace core {
+
+/// A data structure representing a set of tuples (KeyID, flag) in a somewhat
+/// compact form.
+// FIXME: At some point, figure out the optimal representation for this
+// data structure, which is likely to be a lot of the resident memory size.
+class DependencyKeyIDs {
+  // Dependencies to be kept track of.
+  std::vector<KeyID> keys;
+
+  /// Flag indicating if the dependency invalidates the downstream task
+  std::vector<bool> orderOnly;
+  
+  /// Flag indicating if the dependency should not be considered for incremental builds (`true`) or
+  /// if the dependency is valid for incremental builds too (`false`)
+  std::vector<bool> singleUse;
+
+public:
+
+  /// Clear the contents of the set.
+  void clear() {
+    keys.clear();
+    orderOnly.clear();
+    singleUse.clear();
+  }
+  
+  /// Removes entries that are flagged as `singleUse`.
+  void cleanSingleUseDependencies() {
+    bool shouldClean = size() > 0;
+    if (!shouldClean) {
+      return;
+    }
+    
+    std::vector<size_t> indicesToRemove;
+    for (size_t i = 0; i < size(); i++) {
```

Comment:
If you iterate over the vector in reverse here, you can remove the entries as you go without a second loop below

Reply:
Ah right, even better